source/Heuristics.cpp

```diff
@@ -0,0 +1,409 @@
+                       std::size_t k_propagation_max_input_path_leaves)
+    : k_join_override_threshold_(k_join_override_threshold),
+      k_android_join_override_threshold_(k_android_join_override_threshold),
+      k_warn_override_threshold_(k_warn_override_threshold),
+      k_generation_max_port_size_(k_generation_max_port_size),
+      k_generation_max_output_path_leaves_(k_generation_max_output_path_leaves),
+      k_parameter_source_max_port_size_(k_parameter_source_max_port_size),
+      k_parameter_source_max_output_path_leaves_(
+          k_parameter_source_max_output_path_leaves),
+      k_sink_max_port_size_(k_sink_max_port_size),
+      k_sink_max_input_path_leaves_(k_sink_max_input_path_leaves),
+      k_call_effect_source_max_port_size_(k_call_effect_source_max_port_size),
+      k_call_effect_source_max_output_path_leaves_(
+          k_call_effect_source_max_output_path_leaves),
+      k_call_effect_sink_max_port_size_(k_call_effect_sink_max_port_size),
+      k_call_effect_sink_max_input_path_leaves_(
+          k_call_effect_sink_max_input_path_leaves),
+      k_max_number_iterations_(k_max_number_iterations),
+      k_max_depth_class_properties_(k_max_depth_class_properties),
+      k_max_call_chain_source_sink_distance_(
+          k_max_call_chain_source_sink_distance),
+      k_propagation_max_input_path_size_(k_propagation_max_input_path_size),
+      k_propagation_max_input_path_leaves_(
+          k_propagation_max_input_path_leaves) {}
+
+Heuristics Heuristics::from_json(const Json::Value &value) {
+  // Create an `Heuristics` object with the default values.
+  Heuristics heuristics = Heuristics();
+  JsonValidation::validate_object(value);
+  JsonValidation::check_unexpected_members(
+      value,
+      {k_join_override_threshold_attribute,
+       k_android_join_override_threshold_attribute,
+       k_warn_override_threshold_attribute,
+       k_generation_max_port_size_attribute,
+       k_generation_max_output_path_leaves_attribute,
+       k_parameter_source_max_port_size_attribute,
+       k_parameter_source_max_output_path_leaves_attribute,
+       k_sink_max_port_size_attribute, k_sink_max_input_path_leaves_attribute,
+       k_call_effect_source_max_port_size_attribute,
+       k_call_effect_source_max_output_path_leaves_attribute,
+       k_call_effect_sink_max_port_size_attribute,
+       k_call_effect_sink_max_input_path_leaves_attribute,
+       k_max_number_iterations_attribute,
+       k_max_depth_class_properties_attribute,
+       k_max_call_chain_source_sink_distance_attribute,
+       k_propagation_max_input_path_size_attribute,
+       k_propagation_max_input_path_leaves_attribute});
+
+  // Set the heuristics parameters that are specified in the JSON document.
+
+  if (JsonValidation::has_field(value, k_join_override_threshold_attribute)) {
+    heuristics.set_k_join_override_threshold(JsonValidation::unsigned_integer(
+        value, k_join_override_threshold_attribute));
```

Comment:
Having a public `set_` api makes it seem like we can or expect to update the heuristics on the fly during analysis. We should only be able to load the heuristics using the constructor or `Heuristics::from_json()` right? I would suggest removing all the setters and setting them here directly as: 
```
heuristics.k_join_override_threshold_ = JsonValidation::unsigned_integer(value, k_join_override_threshold_attribute);
```

Reply:
Solved in 4c18ae8dcdb20a9edaea79aa40e21792f2610384